Add rx-toggle-switch to framework.

megalord opened the pull request:
Fixes #860 

![screen shot 2015-04-03 at 10 19 35 am](https://cloud.githubusercontent.com/assets/5414922/6984096/f4c10436-d9ea-11e4-92ec-2de4eb52232e.png)

CITguy commented:
I feel like it should it be called `rx-toggle-switch`.  My first reaction to seeing `rx-switch` was that it is an angular implementation of a `switch...case` structure. This is obviously not the case, but `rx-toggle-switch` seems more semantically correct.

glynnis commented:
@megalord Can you also add a demo to the style guide in the form section? (http://rackerlabs.github.io/encore-ui/#/styleguide/forms)

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

halkon commented:
I totally :+1: `rx-toggle-switch` the `rx-switch` name was never really liked and I think `rx-toggle-switch` was preferred for when importing into the framework.

megalord commented:
I agree, `rx-toggle-switch` is a better name; I'll make the change.

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

parlarjb commented on src/rxToggleSwitch/rxToggleSwitch.js:
```diff
@@ -0,0 +1,68 @@
+            model: '=',
+            loading: '=?',
+            disabled: '=?',
+            postHook: '&',
```
This is optional, so let's add the `?`

halkon replied:
Optional I believe only matters for 2 way binding.  
For expressions or attributes, it will just ignore the attribute if not passed (instead of crashing like in the case of a 2-way binding)

https://docs.angularjs.org/api/ng/service/$compile#-scope-

parlarjb commented on src/rxToggleSwitch/rxToggleSwitch.js:
```diff
@@ -0,0 +1,68 @@
+                }
+                var newValue = switchValue(value);
+
+                scope.model = newValue;
```
Instead of having a `.model` attribute, I'd love to just use `ng-model` and the `ngModelController`. With `$setViewValue` and `$render()`, we should be able to get what we'd need, and you'd be able to remove the `model` from the template.

megalord replied:
Good idea! I'll give it a shot.

halkon commented on src/rxToggleSwitch/rxToggleSwitch.js:
```diff
@@ -0,0 +1,68 @@
+            trueValue: '@?',
+            falseValue: '@?',
+            trueText: '@?',
+            falseText: '@?'
```
Given the comment above... We could take off the `?` off of these

megalord replied:
agreed

parlarjb commented:
@glynnis are you ok with the spacing between the right of the knob and the left of the text, when it's in an "off" position?

glynnis commented:
@parlarjb I think there needs to be more spacing there.

Also from a design perspective, I think having a switch that allows for text other than "on/off" seems ripe for misuse (the one that says "Disabled").

The point of a switch is: it's on or off, or cannot be changed. Disabling is communicated by its appearance/color only, not by text in the switch. If the visual treatment for a disabled switch doesn't communicate that clearly enough, I'd rather work on the visual design than allow for text other than "On" and "Off". Opening it up so that devs can put any text there opens up the possibility that someone would use a switch in an incorrect context, like a scenario where a set of radio buttons is more appropriate.

So, basically: can we remove the example and code that allows for the switch that says "Disabled"? 

@megalord let me know if you want me to tweak CSS for spacing and/or work on the visual cues for when a switch is disabled. Happy to let you handle it and post a new screen shot, but I can jump in if you need me.

glynnis commented:
To frame the problem of a switch with text via a silly example:
If you have a light switch in your house, you expect it to control one thing, and to have a state of "on" or "off". Having text on a switch in the UI is kind of like having a light switch that in an "on" position turns on your kitchen light, and in an "off" position turns your dryer on "tumble". 

By allowing text in a switch, we leave ourselves open to the possibility of switches that are doing complex actions that are not true on/off situations, which dilutes the meaning and use of a switch, and would be confusing to the user.

Hope that makes sense.

megalord commented:
Yeah, removing the ability to set the text makes sense.  I also updated the style of the component and added it to the styleguide as suggested.
![screen shot 2015-04-07 at 3 28 20 pm](https://cloud.githubusercontent.com/assets/5414922/7032793/c6c47b6e-dd3a-11e4-9132-a039b56e2db2.png)
Let me know what you think. :smile:

halkon commented:
That's actually a very valid point, and it may have been over-engineered from get go (last year).  The only purpose of being able to customize the text was so that you could change between:

Yes/No
On/Off
True/False

And have the flexibility of doing so (Enable/Disable would fit this obviously).  

But it can get confusing if on every page every switch changes it's true/false visual representation.   Thanks for the update @megalord

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

halkon commented:
I R DONE with DIS PR

![LGTM](https://cloud.githubusercontent.com/assets/1437154/7047675/ad917180-ddd3-11e4-8923-8ed27b5c7375.gif)

LGTM

CITguy commented:
I'll defer to @glynnis on this, but there's not enough contrast between the gray text and gray background for the disabled switch.

glynnis commented:
@megalord and I are gonna sync up about some of the design concerns for this. Adding a "do not merge" for now until we get things rolling on that front.

glynnis commented:
Oops! Didn't mean to close this with my errant click. 

Played around with colors to try to better distinguish disabled from active switches. Also improved the text contrast:

![screen shot 2015-04-08 at 1 17 04 pm](https://cloud.githubusercontent.com/assets/1529366/7052126/28b53fa4-ddf2-11e4-90a9-fbb02341c010.png)

Jordan has the color codes for these changes to implement. Still working on design improvements for the "spinner" switch, so leaving this as "do not merge".

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

CITguy commented:
@glynnis what would a disabled "On" look like?

megalord commented:
@CITguy this is what we settled on:
![screen shot 2015-04-08 at 3 36 55 pm](https://cloud.githubusercontent.com/assets/5414922/7054846/2247c3f4-de05-11e4-92fd-6b26716c0970.png)

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

megalord commented:
I removed the `loading` attribute thanks to @glynnis 's feedback.  I modified the demo to show how a toggle that fires an api call would work.

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

glynnis commented:
Design: LGTM.

@halkon @CITguy Can you guys look this over again now that Jordan has completed changes?

halkon commented:
![LGTM](https://cloud.githubusercontent.com/assets/1437154/7073018/273f1f54-deb6-11e4-98c1-0d783e6b7ddf.gif)

:+1: :+1: :+1: :+1: :+1:

CITguy commented on src/rxToggleSwitch/rxToggleSwitch.less:
```diff
@@ -0,0 +1,92 @@
+
+    -webkit-box-shadow: inset 1px 2px @blurDist @blurSpread rgba(0,0,0,0.50);
+       -moz-box-shadow: inset 1px 2px @blurDist @blurSpread rgba(0,0,0,0.50);
+            box-shadow: inset 1px 2px @blurDist @blurSpread rgba(0,0,0,0.50);
```
You can use the `.box-shadow()` mixin for this.

megalord replied:
Oooh good to know

CITguy commented on src/rxToggleSwitch/rxToggleSwitch.less:
```diff
@@ -0,0 +1,92 @@
+
+        -webkit-box-shadow: 0 0 @blurDist @blurSpread rgba(0,0,0,0.55);
+           -moz-box-shadow: 0 0 @blurDist @blurSpread rgba(0,0,0,0.55);
+                box-shadow: 0 0 @blurDist @blurSpread rgba(0,0,0,0.55);
```
You can use the `.box-shadow()` mixin for this.

CITguy commented on src/rxToggleSwitch/rxToggleSwitch.less:
```diff
@@ -0,0 +1,92 @@
+         -moz-user-select: @select;
+          -ms-user-select: @select;
+              user-select: @select;
+}
```
This should be in the `mixins.less` file. Additionally, you can define it simpler as:

```less
.user-select(...) {
  -prefixed-property: @arguments;
  ...
}
```

megalord replied:
Well, it only takes one argument, so I don't feel like a splat is necessary.  To me, a splat tells the reader that multiple arguments are expected.

CITguy replied:
I tend to use a splat `(...)` when I write a wrapper for vendor-prefixed properties to denote that it accepts the standard arguments, but the functionality is pretty much the same.  I just like keeping things consistent.

CITguy commented on src/rxToggleSwitch/docs/rxToggleSwitch.js:
```diff
@@ -0,0 +1,22 @@
+/*jshint unused:false*/
+
+// This file is used to help build the 'demo' documentation page and should be updated with example code
+function rxToggleSwitchCtrl ($scope, $timeout, rxNotify) {
```
The component template should probably be updated to do this, but we should define the demo controllers using the `.controller('ctrlName', function () {...})` syntax for angular 1.3 compatibility.

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

CITguy commented on src/styles/mixins.less:
```diff
@@ -74,3 +74,12 @@
+    -moz-user-select: @select;
+    -ms-user-select: @select;
+    user-select: @select;
+}
```
I tend to use a splat `(...)` when I write a wrapper for vendor-prefixed properties to denote that it accepts the standard arguments. However, to LESS, the functionality should be the same.

CITguy commented:
Have you rebased?

halkon commented:
Does he particularly need to? The PR is merge-able

CITguy commented:
@halkon: Yes.  I'm seeing an old version of the [`.border-radius()`](https://github.com/rackerlabs/encore-ui/blob/rx-switch/src/styles/mixins.less#L24) mixin that was changed in #888, so I wouldn't want to introduce any unforeseen regressions.

halkon commented:
I think git will handle it just fine, that being said, `.border-radius(...)` is definitely not being used in this PR

halkon commented:
@glynnis https://github.com/rackerlabs/encore-tq-ui/pull/275 is being referenced as using `Yes`/ `No`, mind you while implemented last year, the idea was to cover for these cases. I'll find out (given due time) the reasoning for using `Yes`/ `No` versus `On`/ `Off`, however not to disagree with the decision taken to eliminate the ability to define `On`/ `Off` values, just trying to, rather than cover all cases, make sure we're not missing any

halkon commented:
DOH! posted before some searching in a bigger PR: 

https://github.com/rackerlabs/encore-tq-ui/pull/266/files#diff-dea126b201803ec1dab674699f938bdbR19

Linguistically it would seem more proper to say that "A Feature" is "personal" as a `Yes`/ `No` versus a `On`/ `Off`

May be minimal, and it could probably just be a setting (much like the other PR mentioned) to just change from `On`/ `Off` to a more personal `Yes`/ `No`.

halkon commented:
@glynnis talked with @solim8r today and more or less it came up with, if we're not gonna allow custom text, why even show any text at all?  We could go full mobile mode in the fact that "Green" means on and "Gray" means off.

Throwing that comment out there, I think it would get rid of any linguistics/grammar issues and would likely just make this whole thing simpler.

glynnis commented:
@halkon I think it would be fine to offer an alternate "yes / no" version of the switch, as that keeps with the function and purpose of a switch, as long as the "yes / no" version is hard-coded and not the result of allowing folks to fill in whatever they want.

My problem with alternate text is with the idea of allowing consumers of the switch to add whatever text they want -- that's where it becomes problematic due to potential for long text breaking the design and/or folks filling in text that isn't true to the switch's function and purpose.

I do think it's important that we keep text present on the switch, as there's some general user research that suggests that not all users find switch design as intuitive as others, and the "on/off" or "yes/no" helps communicate its function more clearly.

halkon commented:
Oh I don't disagree with the custom text comment :grin:, I can definitely see the disadvantage on it, and unless we implement a dictionary in it to validate opposites... better to not.

That being said, offering a few other alternatives would be a good thing.  I think Yes/No would be the first, however, it can come in a later PR if need be.  

Interesting about the user research, it makes sense, I kind of was thinking about that this morning on my way in to the office.  "Is it really intuitive that `Green vs Gray` means on/off?".

I think even in essence `Green vs Red` might confuse people even if we deal with it on a day to day basis with stop lights.

@CITguy are we still waiting on the rebase?

megalord commented:
rebased and squashed

halkon commented:
:astonished: :scream: Merge at will!

parlarjb commented on src/rxToggleSwitch/docs/rxToggleSwitch.html:
```diff
@@ -0,0 +1,29 @@
+    <p>
+        <h3>Simple use</h3>
+        <rx-toggle-switch class="demo-default-values" ng-model="encoreSuperpowers"></rx-toggle-switch>
+        <span>Do you have powers? {{ encoreSuperpowers }}</span>
```
Because `encoreSuperpowers` is not initially set on the scope, it goes to `undefined`, and nothing _initially_ shows up here.
![screen shot 2015-04-13 at 5 17 30 pm](https://cloud.githubusercontent.com/assets/9376/7125709/2d44f628-e201-11e4-9c71-648acf764943.png)

Add `$scope.encoreSuperpowers = false` to your `rxToggleSwitchCtrl`, and you'll get:
![screen shot 2015-04-13 at 5 18 03 pm](https://cloud.githubusercontent.com/assets/9376/7125717/46474a90-e201-11e4-8ec2-8f6691c96bd1.png)

megalord replied:
Oh, good point, I guess model needs to be initialized with `false-value` (if not yet set).  This could be a source of confusion, though perhaps effectively beatable with good docs.

parlarjb replied:
Yeah, that's probably the right way to do it.

parlarjb commented on src/rxToggleSwitch/docs/rxToggleSwitch.html:
```diff
@@ -0,0 +1,29 @@
+            true-value="active"
+            false-value="disabled">
+        </rx-toggle-switch>
+        <span>Model value: {{ toggle2 }}</span>
```
Same problem as above, see the screenshots there.

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)

parlarjb commented:
LGTM

TheNando commented on src/rxToggleSwitch/rxToggleSwitch.less:
```diff
@@ -0,0 +1,79 @@
+        text-align: left;
+        .knob {
+            left: 100%;
+            -webkit-transform: translate(-100%, 0);
```
-webkit- but no -moz-?

megalord replied:
Ah, good catch.

TheNando commented on src/rxToggleSwitch/rxToggleSwitch.less:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+    .knob {
+        transition: all 0.1s ease-out;
```
Likewise, if you include support for older browsers with the -browser- selectors, you should include it here as well.

TheNando commented:
I also assess the looks favorably.

comeatmebro commented:
[Wraith UI diff](http://eb4d4842ea70055541c8-b6a3ba129b5eb76af13f4d17e4c1a53b.r51.cf5.rackcdn.com/881/gallery.html)
